internal/infrastructure/kubernetes/ratelimit/resource.go

```diff
@@ -5,10 +5,120 @@
 
 package ratelimit
 
-import "fmt"
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+
+	egcfgv1a1 "github.com/envoyproxy/gateway/api/config/v1alpha1"
+)
+
+const (
+	// RedisSocketTypeEnvVar is the redis socket type.
+	RedisSocketTypeEnvVar = "REDIS_SOCKET_TYPE"
+	// RedisURLEnvVar is the redis url.
+	RedisURLEnvVar = "REDIS_URL"
+	// RuntimeRootEnvVar is the runtime root.
+	RuntimeRootEnvVar = "RUNTIME_ROOT"
+	// RuntimeSubdirectoryEnvVar is the runtime subdirectory.
+	RuntimeSubdirectoryEnvVar = "RUNTIME_SUBDIRECTORY"
+	// RuntimeIgnoreDotfilesEnvVar is the runtime ignoredotfiles.
+	RuntimeIgnoreDotfilesEnvVar = "RUNTIME_IGNOREDOTFILES"
+	// RuntimeWatchRootEnvVar is the runtime watch root.
+	RuntimeWatchRootEnvVar = "RUNTIME_WATCH_ROOT"
+	// LogLevelEnvVar is the log level.
+	LogLevelEnvVar = "LOG_LEVEL"
+	// UseStatsdEnvVar is the use statsd.
+	UseStatsdEnvVar = "USE_STATSD"
+	// InfraName is the name for rate-limit resources.
+	InfraName = "envoy-ratelimit"
+	// InfraGRPCPort is the grpc port that the rate limit service listens on.
+	InfraGRPCPort = 8081
+)
 
 // GetServiceURL returns the URL for the rate limit service.
 // TODO: support custom trust domain
 func GetServiceURL(namespace string) string {
 	return fmt.Sprintf("grpc://%s.%s.svc.cluster.local:%d", InfraName, namespace, InfraGRPCPort)
 }
+
+// rateLimitLabels returns the labels used for all envoy rate limit resources.
+func rateLimitLabels() map[string]string {
+	return map[string]string{
+		"app.gateway.envoyproxy.io/name": InfraName,
+	}
+}
+
+// expectedRateLimitContainers returns expected rateLimit containers.
+func expectedRateLimitContainers(ratelimit *egcfgv1a1.RateLimit, rateLimitDeployment *egcfgv1a1.KubernetesDeploymentSpec) []corev1.Container {
+	ports := []corev1.ContainerPort{
+		{
+			Name:          "http",
+			ContainerPort: InfraGRPCPort,
+			Protocol:      corev1.ProtocolTCP,
+		},
+	}
+
+	containers := []corev1.Container{
+		{
+			Name:            InfraName,
+			Image:           *rateLimitDeployment.Container.Image,
+			ImagePullPolicy: corev1.PullIfNotPresent,
+			Command: []string{
+				"/bin/ratelimit",
+			},
+			Env:             expectedRateLimitContainerEnv(ratelimit, rateLimitDeployment),
+			Ports:           ports,
+			Resources:       *rateLimitDeployment.Container.Resources,
+			SecurityContext: rateLimitDeployment.Container.SecurityContext,
+			VolumeMounts: []corev1.VolumeMount{
+				{
+					Name:      InfraName,
+					MountPath: "/data/ratelimit/config",
+					ReadOnly:  true,
+				},
+			},
+			TerminationMessagePolicy: corev1.TerminationMessageReadFile,
+			TerminationMessagePath:   "/dev/termination-log",
+		},
+	}
+
+	return containers
+}
+
+// expectedRateLimitContainerEnv returns expected ratelimit container envs.
+func expectedRateLimitContainerEnv(ratelimit *egcfgv1a1.RateLimit, rateLimitDeployment *egcfgv1a1.KubernetesDeploymentSpec) []corev1.EnvVar {
+	env := []corev1.EnvVar{}
+
+	defaultEnvMapping := map[string]string{
```

Comment:
the issue is with this logic, I think we should stay away from `maps`, the number of elements here is small,
so the improvement in time complexity is minimal.

I propose this logic
* add default env vars into `env` list
* iterate through user defined env var
    * find and replace if exists
* lets not do anything special like skip `RedisURLEnvVar`, if the user
wants to override it, we can let them, they are on their own

Reply:
When the user setting more env than default env, we should merge these env, override default env by the user setting and keep another default env, imo if the user can override default env , the default env is unnecessary in eg and use the user setting env override `ratelimit` default env settings.

Reply:
* so reiterating that using a `map` is not a good idea, ordering within a map is unstable, so every time the map is translated into a list, the order may change causing churn in the K8s API server and will probably also end up restarting the envoy pod
* regarding overriding backend settings - the backend setting is the common case to enable rate limiting, editing the `env` is an edge case which very few users might configure, and if they choose to override ` RedisURLEnvVar`, we should probably let them because that is the user intent. In the future, if we want to add guardrails, we should consider adding validation early on in the pipeline during EG startup

Reply:
done, pls review again @arkodg